### PR TITLE
Array aware call cse

### PR DESCRIPTION
**Pull request by arrangabriel**
Implements array aware CSE for matching iteration spaces. 
Ref issue: #4

### CODE REVIEW

File: lib/Dialect/BaseModelica/Transforms/CallCSE.cpp [hunk 1]
```diff
@@ -17,80 +19,103 @@ class CallCSEPass final : public impl::CallCSEPassBase<CallCSEPass> {
...
+
+/// Reduces instances to a single one per template.
+/// Completely filters out instances where induction variable ranges do not
+/// match.
+llvm::SmallVector<EquationInstanceOp>
+filterInstanceOps(llvm::SmallVectorImpl<EquationInstanceOp> &instanceOps) {
+  llvm::SmallVector<EquationInstanceOp> filteredInstanceOps;
+  llvm::DenseMap<EquationTemplateOp, llvm::SmallVector<EquationInstanceOp>>
+      instancesByTemplate;
+  for (EquationInstanceOp equationOp : instanceOps) {
```
Comment (mscuttari) [inline lib/Dialect/BaseModelica/Transforms/CallCSE.cpp:46]: You can ask for the users of the `equation_template` result value.

File: lib/Dialect/BaseModelica/Transforms/CallCSE.cpp [hunk 2]
```diff
@@ -17,80 +19,103 @@ class CallCSEPass final : public impl::CallCSEPassBase<CallCSEPass> {
...
 };
 
-/// Get all call operations in the model.
-void collectCallOps(ModelOp modelOp, llvm::SmallVectorImpl<CallOp> &callOps) {
+bool haveSameInductionRanges(
+    llvm::SmallVectorImpl<EquationInstanceOp> &instanceOps) {
+  std::optional<IndexSet> seenSpace;
+  for (EquationInstanceOp instanceOp : instanceOps) {
+    IndexSet iterationSpace = instanceOp.getIterationSpace();
+    if (!seenSpace.has_value()) {
```
Comment (mscuttari) [inline lib/Dialect/BaseModelica/Transforms/CallCSE.cpp:29]: `optional` has implicit cast to `bool`. No need for `has_value`.

File: lib/Dialect/BaseModelica/Transforms/CallCSE.cpp [hunk 3]
```diff
@@ -17,80 +19,103 @@ class CallCSEPass final : public impl::CallCSEPassBase<CallCSEPass> {
...
-               DynamicOp dynamicOp, mlir::SymbolTable &symbolTable,
-               mlir::RewriterBase &rewriter);
 };
 
-/// Get all call operations in the model.
-void collectCallOps(ModelOp modelOp, llvm::SmallVectorImpl<CallOp> &callOps) {
+bool haveSameInductionRanges(
+    llvm::SmallVectorImpl<EquationInstanceOp> &instanceOps) {
+  std::optional<IndexSet> seenSpace;
+  for (EquationInstanceOp instanceOp : instanceOps) {
```
Comment (mscuttari) [inline lib/Dialect/BaseModelica/Transforms/CallCSE.cpp:27]: Check if you have more than one instance. If not, there's no need for checking the iteration space.

File: lib/Dialect/BaseModelica/Transforms/CallCSE.cpp [hunk 4]
```diff
@@ -17,80 +19,103 @@ class CallCSEPass final : public impl::CallCSEPassBase<CallCSEPass> {
...
+        instancesByTemplate.getOrInsertDefault(equationOp.getTemplate());
+    group.push_back(equationOp);
+  }
+
+  for (auto [_, instanceOps] : instancesByTemplate) {
+    if (haveSameInductionRanges(instanceOps)) {
+      filteredInstanceOps.push_back(instanceOps.front());
+    }
+  }
+  llvm::stable_sort(filteredInstanceOps);
```
Comment (mscuttari) [inline lib/Dialect/BaseModelica/Transforms/CallCSE.cpp:57]: Why do you need this?

File: lib/Dialect/BaseModelica/Transforms/CallCSE.cpp [hunk 5]
```diff
@@ -17,80 +19,103 @@ class CallCSEPass final : public impl::CallCSEPassBase<CallCSEPass> {
...
-    llvm::SmallVector<CallOp> *equivalenceGroup = find_if(
-        tmpCallEquivalenceGroups, [&](llvm::SmallVector<CallOp> &group) {
+llvm::SmallVector<llvm::SmallVector<CallWithIndices>>
+buildEquivalenceGroups(const llvm::SmallVectorImpl<CallWithIndices> &callOps) {
+  llvm::SmallVector<llvm::SmallVector<CallWithIndices>> callEquivalenceGroups;
+  for (CallWithIndices callOp : callOps) {
+    auto callExpression = mlir::dyn_cast<EquationExpressionOpInterface>(
+        callOp.first.getOperation());
+
+    mlir::SymbolTableCollection symbolTableCollection;
```
Comment (mscuttari) [inline lib/Dialect/BaseModelica/Transforms/CallCSE.cpp:96]: Do not recreate a `SymbolTableCollection` every time. This should be create only once within the pass.

File: lib/Dialect/BaseModelica/Transforms/CallCSE.cpp [hunk 6]
```diff
@@ -126,96 +151,283 @@ mlir::Operation *cloneDefUseChain(mlir::Operation *op,
...
-void CallCSEPass::emitCse(llvm::SmallVectorImpl<CallOp> &equivalenceGroup,
-                          ModelOp modelOp, DynamicOp dynamicOp,
-                          mlir::SymbolTable &symbolTable,
-                          mlir::RewriterBase &rewriter) {
-  assert(!equivalenceGroup.empty() && "equivalenceGroup cannot be empty");
-  CallOp representative = equivalenceGroup.front();
-  const mlir::Location loc = representative.getLoc();
+/// Get the indices of induction variables accessed by [callOp].
+llvm::SmallVector<int64_t>
+getAccessedIndices(CallOp callOp, MultidimensionalRange &indices,
```
Comment (mscuttari) [inline lib/Dialect/BaseModelica/Transforms/CallCSE.cpp:167]: This function needs much more documentation, as it seems to be core to the solution. It is not clear whose indices are accessed. Does "accessed indices" mean "used induction variables"? Or "accessed equation indices"? Or maybe else? The initial comment seems to suggest the latter, but the former is then what actually happens.

File: lib/Dialect/BaseModelica/Transforms/CallCSE.cpp [hunk 7]
```diff
@@ -126,96 +151,283 @@ mlir::Operation *cloneDefUseChain(mlir::Operation *op,
...
+          accessedIndices.push_back(blockArg.getArgNumber());
+        }
+      }
+    }
+  }
+
+  llvm::sort(accessedIndices);
+
+  llvm::SmallVector<int64_t> indexMapping;
+  // We need to sort the index mapping to match the baseMapping
```
Comment (mscuttari) [inline lib/Dialect/BaseModelica/Transforms/CallCSE.cpp:190]: What is `baseMapping`?

File: lib/Dialect/BaseModelica/Transforms/CallCSE.cpp [hunk 8]
```diff
@@ -126,96 +151,283 @@ mlir::Operation *cloneDefUseChain(mlir::Operation *op,
...
-                          mlir::SymbolTable &symbolTable,
-                          mlir::RewriterBase &rewriter) {
-  assert(!equivalenceGroup.empty() && "equivalenceGroup cannot be empty");
-  CallOp representative = equivalenceGroup.front();
-  const mlir::Location loc = representative.getLoc();
+/// Get the indices of induction variables accessed by [callOp].
+llvm::SmallVector<int64_t>
+getAccessedIndices(CallOp callOp, MultidimensionalRange &indices,
+                   const std::optional<MultidimensionalRange> &baseMapping) {
+  llvm::SmallVector<int64_t> accessedIndices;
```
Comment (mscuttari) [inline lib/Dialect/BaseModelica/Transforms/CallCSE.cpp:169]: Why not a set (`llvm::DenseSet`)? Also, if the intention is to collect the used inductions, the name `accessedIndices` is misleading, resembling the actual indices of the iteration space. `usedInductions` would be a better name.

File: lib/Dialect/BaseModelica/Transforms/CallCSE.cpp (hunk 7 above)
Comment (mscuttari) [inline lib/Dialect/BaseModelica/Transforms/CallCSE.cpp:190]: Not clear what you're trying to do.

**Review by mscuttari — CHANGES_REQUESTED** [review]
I have left some comments, but I have also stopped reviewing after the `getAccessedIndices` function. The overall impression is that much more documentation is needed, as the current implementation is quite obscure and requires too much reverse engineering to be understood. As a result of this I have the feeling, which may wrong, that you're trying to analyze the access functions and somehow reorder the induction variables later. This is a very complicated task which has been already encountered, and many more examples should be used to comment the code intentions.

File: lib/Dialect/BaseModelica/Transforms/CallCSE.cpp (hunk 1 above)
Comment (arrangabriel) [reply lib/Dialect/BaseModelica/Transforms/CallCSE.cpp:46]: I'm not sure I understand how that would help here, as I already have all the instances, and just want to group them by template

Comment (mscuttari) [reply lib/Dialect/BaseModelica/Transforms/CallCSE.cpp:46]: True

**Comment by arrangabriel** [comment]
Yeah, it's definitely a bit of a mess. I'm rewriting it to simplify now, I'll let you know when you can take a look at it again. I'll also add some more comments to describe the tricky parts.

File: lib/Dialect/BaseModelica/Transforms/CallCSE.cpp (hunk 4 above)
Comment (mscuttari) [reply lib/Dialect/BaseModelica/Transforms/CallCSE.cpp:57]: Please add examples to clarify. Also it's still unclear to me what that means. You have a list of instances, and you're sorting them by the address of the underlying `mlir::Operation`.

**Comment by arrangabriel** [comment]
@mscuttari I made some major changes. Hopefully the intentions behind the code are clearer now, helped by some more comments as well. Let me know what you think.

The thing that makes the whole thing somewhat tricky is this: After the cse variable has been created, along with an equation to drive its value the calls have to be replaced with accesses to it. In the array case the cse tensor must be indexed with the induction variables of the equation that contained the call. It is not a one-to-one replacement though, as I wanted to support calls only using a subset of the induction variables, and also using them out of order, so long as the iteration spaces match. So all the somewhat obtuse code refering to accesses and inductions handles this remapping. To simplify it I used the observation that one doesn't have to use the "correct" variable, as long as the ranges match. The `matchInductionOrder` function now just looks through the needed ranges, and grabs them from the available inductions in the parent template of the call we are replacing.

**Comment by mscuttari** [comment]
If I understand correctly what you're say, the last hypothesis is unfortunately wrong.
A possible counterexample is the following:
```modelica
for i in 1:10 loop
  for j in 1:10 loop
    x = foo(i, j) ...;
  end for;
end for;
```
becoming
```modelica
for i in 1:10 loop
  for j in 1:10 loop
    cse = foo(i, j) ...;
    x = cse[i, j];
  end for;
end for;
```
In the access to `cse`, you can't replace `j` with `i` just because their range is the same.
I will look at the code tomorrow.

**Comment by arrangabriel** [comment]
I think you may be right about the first part, that's annoying. I did have another approach in mind, but it's not particularly elegant. 

The order could be decided by the traversal of the def-use chain for each operation. Because they are necessarily the same for two that were marked as equivalent. I don't love this approach though, because if the isEquivalent interface would change to for example handle symmetric operations it would break.

As for the second part I'm not quite sure what you mean.
It would go from:
```
loop i 1:5
  loop j 1:10
    x1 = foo(i / j)

loop i 1:10
  loop j 1:5
    x2 = foo(j / i)
``` 
To 
```
loop i 1:5
  loop j 1:10
    x1 = cse[i, j]

loop i 1:10
  loop j 1:5
    x2 = cse[j, i]

loop i 1:5
  loop j 1:10
    cse[i, j] = foo(i / j)
```
(Changed it to a division so the order matters)

Unless I misunderstood what problem you were getting at.

**Comment by mscuttari** [comment]
I see what you mean with the use-def chains, and I share the same view about the consequent risks for the interface. Even if probably working in this initial version, I personally wouldn't go that route, keeping things simple is key for future development. What if we just limit to have the same iteration space? After all, I expect this to be the most common case. And even if it's not, we can still add another pass (or a sub-transformation within this one) that make the iteration loops as similar as possible.
For the second part I think you might be right, it was late and I got a bit confused.

**Comment by mscuttari** [comment]
The general rule is to keep everything as simple as possible. The applicability of this optimization will be quite limited anyway.
This being said, this is what I believe to be the best solution:
 - Remove the iteration ranges from the interfaces (roll it back to the original version), it's just an over-complication (e.g., operations like `ReductionOp` may add additional dimensions).
 - Perform the tree traversal and determine the order of the used induction variables. Keep in mind that the traversal may encounter operations with regions.
 - If the order of induction variables for two different calls, together with their associated ranges, match, then the calls can be CSEd. It is theoretically possible to CSE only the minimum common iteration space, like it is done for derivatives materialization, but this involves adding additional equations for filling the gaps. I would keep it for a later moment, there's no need for it now (and, imho, there will probably never be).

P.S.: as for the benchmarks, I will leave the optimization last, as we will need it to be merged to master before the deadline to have a fixed commit to be reported.

**Comment by arrangabriel** [comment]
Yep, I'm working on something pretty close to what you describe here.

**Comment by arrangabriel** [comment]
I made some major changes now @mscuttari . 

As you said I removed the indices from the isEquivalent call. Instead I have it build up a mapping between induction variables accessed by each operation. This is so we later on can compare the ranges of the variables they access, and access the cse variable correctly.

I put more effort into commenting what the various functions do, hopefully it makes things more clear (in combination with the approach being a lot less convoluted imo).

**Comment by mscuttari** [comment]
I'm a bit against having that `SmallVector` as argument to `isEquivalent`. 
A separate function `getInductionVisits(SmallVectorImpl&)` would be better, given that this piece of information is only needed to this pass. I know that a double traversal would be performed, but I find it cleaner from the code perspective. You can also provide a default implementation recursively visiting the operands and their defining operations (see the `getAccesses` method to better understand what I mean).
I will have a look at the rest of the code tomorrow morning as soon as I wake up!

**Comment by arrangabriel** [comment]
I somewhat agree, which is why I didn't take this approach initially, but I convinced myself this was a good way to do it for two reasons.

1. Comparing the inductions variables that are read can be seen as necessary to check if two expressions are equivalent. This at least is definitely the case for this pass. If one does not care about it one can just ignore the output, or check that it is empty if inductions should be treated as non-equivalent.

2. It's somewhat of a premature generalization, but if we wanted to at some point support symmetric operations being equivalent `a + b == b + a` I'm not sure how to reconcile that with a separate traversal function. As it would be closely tied to the isEquivalent implementation itself.

Let me know what you think. In any case I don't know if I like the argument name, or how it is explained in the tablegen file, I couldn't immediately come up with anything better though.

**Comment by mscuttari** [comment]
Ok, it makes sense. I agree with the rest of your analysis, apart for a small detail: go for the additional method without the vector argument, and have a default implementation for that which allocates one and call the already existing method expecting it. I prefer this solution to `optional` because, in that case, you would also have to wrap it in a `reference_wrapper`. Moreover, as you said, it is better to have the vector always populated and, if not necessary, just discarded by the caller.

**Comment by arrangabriel** [comment]
Okay, I'll do that then.

The last part confused me a little bit. What you're saying is the overload with the vector treats all inductions as equivalent, and it's the responsibility of the caller to check the actual inductions. Whereas for the overload without the vector it's allocated inside the method, which uses the other overload internally (i.e. treats inductions as equivalent). Then what you mean by syntactically compare is to check if all accessed inductions have the same position?

**Comment by arrangabriel** [comment]
I tried to add the default implementation, but it seems OpInterfaces don't allow method overloading. 
This is the generated code:
image

Should I just create two separate names? If so do you have a preference on the naming?
And do you know how I can use one interface method inside another, I'm struggling a bit to figure it out.
This was my initial attempt, but the call to isEquivalent isn't correct.

```
InterfaceMethod<
    "Check if two expressions are equivalent. Induction accesses are treated as equal, and are returned in [inductionAccesses]",
    "bool", "isEquivalent",
    (ins "::mlir::Operation*":$other,
         "::llvm::SmallVectorImpl<std::pair<::mlir::BlockArgument, ::mlir::BlockArgument>>&":$inductionAccesses,
         "::mlir::SymbolTableCollection&":$symbolTableCollection), "", [{
        // Safely assume that the two expressions are different.
        return false;
    }]>,
InterfaceMethod<
    "Check if two expressions are equivalent. Induction accesses with the same position are treated as equal.",
    "bool", "isEquivalentDefault",
    (ins "::mlir::Operation*":$other,
         "::mlir::SymbolTableCollection&":$symbolTableCollection), [{
         ::llvm::SmallVector<std::pair<mlir::BlockArgument, mlir::BlockArgument>> inductionAccesses;
         bool equivalent = isEquivalent(other, inductionAccesses, symbolTableCollection);
         return equivalent && llvm::all_of(inductionAccesses, [&](auto accessPair) {
             return accessPair.first.getArgNumber() == accessPair.second.getArgNumber();
         });
    }]>,
```

File: lib/Dialect/BaseModelica/Transforms/CallCSE.cpp [hunk 9]
```diff
@@ -117,105 +261,223 @@ mlir::Operation *cloneDefUseChain(mlir::Operation *op,
...
+  rewriter.setInsertionPointToStart(
+      cseEquationTemplateOp.createBody(numInductions));
+
+  // Build equation left hand side by getting the cse variable reference.
+  auto cseVarRef = rewriter.create<VariableGetOp>(loc, cseVariable);
+  EquationSideOp lhsOp;
+  if (numInductions == 0) {
+    lhsOp = rewriter.create<EquationSideOp>(loc, cseVarRef->getResults());
+  } else {
+    // Extract cse tensor slot with template inductions.
```
Comment (mscuttari) [inline lib/Dialect/BaseModelica/Transforms/CallCSE.cpp:303]: I'm not sure this is correct.
Let's say the original equation range was [5,8]. The CSE variable will have 4 elements. The driving equation instance will have range [5,8], and will be the only usage of the driving template. Within the template, the accesses to the CSE variable should be shifted back by 5 positions, not 1. Or am I missing something? The other usages of the inductions (i.e., within the function call) should instead remain untouched.

**Review by mscuttari — CHANGES_REQUESTED** [review]

**Comment by mscuttari** [comment]
> I tried to add the default implementation, but it seems OpInterfaces don't allow method overloading. This is the generated code: image
> 
> Should I just create two separate names? If so do you have a preference on the naming? And do you know how I can use one interface method inside another, I'm struggling a bit to figure it out. This was my initial attempt, but the call to isEquivalent isn't correct.
> 
> ```
> InterfaceMethod<
>     "Check if two expressions are equivalent. Induction accesses are treated as equal, and are returned in [inductionAccesses]",
>     "bool", "isEquivalent",
>     (ins "::mlir::Operation*":$other,
>          "::llvm::SmallVectorImpl<std::pair<::mlir::BlockArgument, ::mlir::BlockArgument>>&":$inductionAccesses,
>          "::mlir::SymbolTableCollection&":$symbolTableCollection), "", [{
>         // Safely assume that the two expressions are different.
>         return false;
>     }]>,
> InterfaceMethod<
>     "Check if two expressions are equivalent. Induction accesses with the same position are treated as equal.",
>     "bool", "isEquivalentDefault",
>     (ins "::mlir::Operation*":$other,
>          "::mlir::SymbolTableCollection&":$symbolTableCollection), [{
>          ::llvm::SmallVector<std::pair<mlir::BlockArgument, mlir::BlockArgument>> inductionAccesses;
>          bool equivalent = isEquivalent(other, inductionAccesses, symbolTableCollection);
>          return equivalent && llvm::all_of(inductionAccesses, [&](auto accessPair) {
>              return accessPair.first.getArgNumber() == accessPair.second.getArgNumber();
>          });
>     }]>,
> ```

Oh well, that's annoying, I was forgetting how MLIR does things underneath. At the moment I don't have any name preference, feel free to decide it.
For calling the other method, have you tried casting `$_op.getOperation()` to the interface and call it from there? Like in the first lines of `getEquationAccesses`.

File: lib/Dialect/BaseModelica/Transforms/CallCSE.cpp (hunk 9 above)
Comment (arrangabriel) [reply lib/Dialect/BaseModelica/Transforms/CallCSE.cpp:303]: I'll think about it for a second, but I think you're right.

**Comment by mscuttari** [comment]
Yes

File: lib/Dialect/BaseModelica/Transforms/CallCSE.cpp (hunk 9 above)
Comment (arrangabriel) [reply lib/Dialect/BaseModelica/Transforms/CallCSE.cpp]: I fixed it now, and added a test.

File: test/Dialect/BaseModelica/Transforms/CallCSE/idempotent-configurations.mlir [hunk 10]
```diff
@@ -152,4 +95,72 @@ module @ArrayResult {
...
+            %6 = bmodelica.add %arg1, %4 : (index, index) -> index
+            %7 = bmodelica.variable_get @b : tensor<4x5xi32>
+            %8 = bmodelica.tensor_extract %7[%5, %6] : tensor<4x5xi32>
+            %9 = bmodelica.equation_side %8 : tuple<i32>
+            %10 = bmodelica.call @f(%arg0, %arg1) : (index, index) -> i32
+            %11 = bmodelica.equation_side %10 : tuple<i32>
+            bmodelica.equation_sides %9, %11 : tuple<i32>, tuple<i32>
+        }
+
+        bmodelica.dynamic {
```
Comment (mscuttari) [inline test/Dialect/BaseModelica/Transforms/CallCSE/idempotent-configurations.mlir:160]: Check if the instances have remained the same.

**Review by mscuttari — CHANGES_REQUESTED** [review]
I think it should be ok now. Address the last comment I've left, rebase (commit "Add array-aware function calls CSE"), and I will merge.

**Comment by arrangabriel** [comment]
Nice, I'll let you know when it's done

**Comment by arrangabriel** [comment]
Should be good now @mscuttari

**Review by mscuttari — APPROVED** [review]

**Pull request merged**